compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -2515,6 +2515,28 @@ impl<'tcx> TyCtxt<'tcx> {
         (ident, scope)
     }
 
+    /// Returns `true` if `span` originates in a macro's expansion where debuginfo should be
+    /// collapsed.
+    pub fn in_macro_expansion_with_collapse_debuginfo(
+        self,
+        span: Span,
+        hdata: &HygieneData,
+    ) -> bool {
+        if span.from_expansion() {
+            let outer = hdata.expn_data(hdata.outer_expn(span.ctxt())).clone();
+            let in_std_macro = match outer.macro_def_id {
+                Some(macro_def_id) => {
+                    let crate_name = self.crate_name(macro_def_id.krate);
+                    crate_name == sym::std || crate_name == sym::core
```

Comment:
This misses liballoc and I don't think we should match on the exact crate name.

Reply:
Changed to `[sym::core, sym::alloc, sym::std].contains(&crate_name)`. I didn't find other options to detect if some macro belongs to system crate. Found some cases like this in code:
https://github.com/rust-lang/rust/blob/2fdd9eda0ce4fa0ecbf3099783f4f505235ceb44/compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs#L757
https://github.com/rust-lang/rust/blob/2fdd9eda0ce4fa0ecbf3099783f4f505235ceb44/compiler/rustc_expand/src/config.rs#L106
https://github.com/rust-lang/rust/blob/2fdd9eda0ce4fa0ecbf3099783f4f505235ceb44/src/tools/clippy/clippy_utils/src/eager_or_lazy.rs#L62

Could you recommend some other way?

Reply:
Hardcoding these crates in the compiler is not a good approach in general.
It's better to add the necessary `#[collapse_debuginfo]` attributes explicitly in `library`.

Reply:
Adding `#[collapse_debuginfo]` for standart library macroses requires enabling the corresponding unstable feature for those libraries. How to organize transition period there?

Reply:
`#[cfg_attr(not(bootstrap), collapse_debuginfo]` can be used to only enable the attribute when not compiling using the bootstrap compiler.

Reply:
`collapse_debuginfo` was implemented quite some time ago (https://github.com/rust-lang/rust/pull/99556), so it's available in bootstrap compiler and `cfg_attr` shouldn't be necessary.